File: mns-core/src/main/java/cz/fb/manaus/core/model/PriceComparator.java
package cz.fb.manaus.core.model;

import java.util.Comparator;

import static com.google.common.collect.ComparisonChain.start;

public class PriceComparator implements Comparator<Price> {

    public static final PriceComparator CMP = new PriceComparator();

    @Override
    public int compare(Price o1, Price o2) {
        if (o1.getSide() == o2.getSide()) {
            switch (o1.getSide()) {
                case BACK:
                    return start().compare(o2.getPrice(), o1.getPrice()).compare(o2.getAmount(), o1.getAmount()).result();
                case LAY:
                    return start().compare(o1.getPrice(), o2.getPrice()).compare(o1.getAmount(), o2.getAmount()).result();
                default:
                    throw new IllegalStateException();

            }
        }
        return start().compare(o1.getSide(), o2.getSide()).result();
    }

}

File: mns-repository/src/main/java/cz/fb/manaus/repository/domain/PriceComparator.kt
package cz.fb.manaus.repository.domain


object PriceComparator : Comparator<Price> {

    override fun compare(p1: Price, p2: Price): Int {
        if (p1.side === p2.side) {
            return when (p1.side) {
                Side.BACK -> compareValuesBy(p2, p1, Price::price, Price::amount)
                Side.LAY -> compareValuesBy(p1, p2, Price::price, Price::amount)
            }
        }
        return compareValuesBy(p1, p2, Price::side)
    }
}
